app/services/RestaurantService.java

```diff
@@ -1,157 +1,208 @@
-				break;
-		}
+	List<Restaurant> restaurants = criteria.list();
 
-		return PaginationAdapter.createOutput()
-				.setPageNumber(restaurantFilter.pageNumber)
-				.setPageSize(restaurantFilter.pageSize)
-				.setModel(restaurants)
-				.setNumberOfPages(numberOfPages);
+	switch (restaurantFilter.sortBy)
+	{
+	case "rating":
+	    restaurants.sort((o1, o2) -> o2.getAverageRating().compareTo(o1.getAverageRating()));
+	    break;
 	}
 
-	/**
-	 * Gets restaurant with id.
-	 *
-	 * @param id the id
-	 * @return the restaurant with id
-	 */
-	public Restaurant getRestaurantWithId(final UUID id) {
-		return (Restaurant) getSession().createCriteria(Restaurant.class)
-				.add(Restrictions.eq("id", id))
-				.uniqueResult();
+	return PaginationAdapter.createOutput().setPageNumber(restaurantFilter.pageNumber)
+		.setPageSize(restaurantFilter.pageSize).setModel(restaurants).setNumberOfPages(numberOfPages);
+    }
+
+    /**
+     * Gets restaurant with id.
+     *
+     * @param id
+     *            the id
+     * @return the restaurant with id
+     */
+    public Restaurant getRestaurantWithId(final UUID id)
+    {
+	return (Restaurant) getSession().createCriteria(Restaurant.class).add(Restrictions.eq("id", id)).uniqueResult();
+    }
+
+    /**
+     * Gets nearby restaurants.
+     *
+     * @param latitude
+     *            the latitude
+     * @param longitude
+     *            the longitude
+     * @return the nearby restaurants
+     */
+    @SuppressWarnings("unchecked")
+    public List<Restaurant> getNearbyRestaurants(final Float latitude, final Float longitude)
+    {
+	return getSession()
+		.createSQLQuery(
+			"SELECT * FROM restaurant WHERE restaurant.longitude <> 0 AND restaurant.latitude <> 0 ORDER BY ST_Distance(ST_GeomFromText('POINT(' || restaurant.longitude || ' ' || restaurant.latitude || ')' ,4326), ST_GeomFromText('POINT(' || :longitude || ' ' || :latitude || ')',4326)) ASC LIMIT 3")
+		.addEntity(Restaurant.class).setParameter("longitude", longitude).setParameter("latitude", latitude)
+		.list();
+    }
+
+    /**
+     * Gets popular restaurants.
+     *
+     * @return the popular restaurants
+     */
+    @SuppressWarnings("unchecked")
+    public List<Restaurant> getPopularRestaurants(UUID userId)
+    {
+
+	UserHabitsBean userHabits = getUserHabits(userId);
+      
+	List<Restaurant> restaurants = (List<Restaurant>) getSession().createCriteria(Restaurant.class)
+		.add(Restrictions.eq("city.id", userHabits.getCity()))
+		.add(Restrictions.eq("priceRange", userHabits.getUserPriceRange())).list();
+
+	List<Object> toRemove = new ArrayList<Object>();
+
+	for (Restaurant restaurant : restaurants)
+	{
+
+	    if (restaurant.getAverageRating() < (userHabits.getUserRatingRange() - 0.5d)
+		    || restaurant.getAverageRating() > (userHabits.getUserRatingRange() + 0.5d))
+	    {
+		toRemove.add(restaurant);
+	    }
+
 	}
+	restaurants.removeAll(toRemove);
```

Comment:
Now, that we use pagination, we can not just filter the results after the fetching has been completed, because it will cause a non-uniform distribution per page.